Adding var/list/atom_colours to /atom .

phil235 opened the pull request:
:cl: phil235
add: Spray cleaner and soap can now wash off paint color.
/:cl:

The var will be used to store the various coloring that happen for the atom so that we can separate paint coloring from color that must be inherent to the atom (an initial color for example), or from certain coloring effect like revenant's blight, mob electrocution's black color, admin edit of the color var, green color from holding the greentext item, etc.
The list has four elements, used for four categories:
ADMIN_COLOUR_PRIORITY for admin varedits and very rate color effect like holding the greentext item (and other effects that should prime over any other potential source of coloring even temporary effects).
TEMPORARY_COLOUR_PRIORITY for short color effects like revenant blight on mob, mob electrocution making you all black for a couple seconds, effects that should be appearing above paint coloring.
WASHABLE_COLOUR_PRIORITY for pretty much all paint coloring like colorful reagent on mobs, coloring turfs with paint, etc.
FIXED_COLOUR_PRIORITY for color inherent to the atom, like a blob's color, any object with a color value given in its definition.

Fixes electocution animation on mob not making the mob all black (with the skeleton overlay blinking over it)
Spray cleaner and soap can now wash paint off mobs, turfs and objects. Fixes #7640.

phil235 committed:
Adding var/list/atom_colours to /atom .
The var will be used to store the various coloring that happen for the atom so that we can separate paint coloring from color that must be inherent to the atom (an initial color for example), or from certain coloring effect like revenant's blight, mob electrocution's black color, admin edit of the color var, green color from holding the greentext item, etc.
The list has four elements, used for four categories:
ADMIN_COLOUR_PRIORITY for admin varedits and very rate color effect like holding the greentext item (and other effects that should prime over any other potential source of coloring even temporary effects).
TEMPORARY_COLOUR_PRIORITY for short color effects like revenant blight on mob, mob electrocution making you all black for a couple seconds, effects that should be appearing above paint coloring.
WASHABLE_COLOUR_PRIORITY for pretty much all paint coloring like colorful reagent on mobs, coloring turfs with paint, etc.
FIXED_COLOUR_PRIORITY for color inherent to the atom, like a blob's color, any object with a color value given in its definition.

Fixes electocution animation on mob not making the mob all black (with the skeleton overlay blinking over it)
Spray cleaner and soap can now wash paint off mobs, turfs and objects.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into Color_washing

# Conflicts:
#	code/game/atoms.dm
#	code/game/gamemodes/nuclear/nuclearbomb.dm

ChangelingRain commented on code/game/gamemodes/blob/blobs/core.dm:
```diff
@@ -32,11 +32,7 @@
-	add_overlay(I)
+	..()
 	var/image/C = new('icons/mob/blob.dmi', "blob_core_overlay")
 	add_overlay(C)
```
Image might need RESET_COLOR or something so the overlay isn't colored along with the blob it's on

phil235 commented on code/game/gamemodes/clock_cult/clock_scripture.dm:
```diff
@@ -1135,15 +1136,15 @@ Judgement: 10 servants, 100 CV, and any existing AIs are converted or destroyed
 	playsound(invoker, 'sound/magic/clockwork/invoke_general.ogg', 50, 0)
 	playsound(invoker, 'sound/magic/lightning_chargeup.ogg', 100, 0)
-	animate(invoker, color = list(rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(0,0,0)), time = 88) //Gradual advancement to extreme brightness
+	animate(invoker, color = "#FFFF00", time = 88) //Gradual advancement to extreme brightness
```
using a color matrix in a animate() does not work, the invoker's color is instantly changed to the end color instead of gradually, so I changed it to yellow (because you can't do white with just color text)

ChangelingRain replied:
I'm pretty sure there's a big difference between these two lines; this will not turn the user bright white, only yellow. Yellow looks awful.

ChangelingRain replied:
I can't replicate that failure in testing, it may be a bug with the latest beta or something.

RemieRichards replied:
White is #FFFFFF Phil.

MrPerson replied:
I was also unable to replicate this issue

phil235 replied:
"#FFFFFF" would appear completely transparent though, that's why I opted for yellow.
I'm indeed using the beta version, I guess I'll revert my code.

ChangelingRain commented on code/game/gamemodes/blob/blobs/node.dm:
```diff
@@ -20,13 +20,9 @@
+	..()
 	var/image/C = new('icons/mob/blob.dmi', "blob_node_overlay")
-	src.add_overlay(C)
+	add_overlay(C)
```
Same note on RESET_COLOR

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_scripture.dm:
```diff
@@ -1135,15 +1136,15 @@ Judgement: 10 servants, 100 CV, and any existing AIs are converted or destroyed
 		playsound(invoker, 'sound/magic/lightningbolt.ogg', 100, 0)
 		if(invoker.stat == CONSCIOUS)
-			animate(invoker, color = initial(invoker.color), time = 10)
+			invoker.update_atom_colour()
```
This was a brief animate so that it'd roughly match the lightning bolts.

phil235 commented on code/game/gamemodes/blob/blobs/core.dm:
```diff
@@ -32,11 +32,7 @@
-	var/image/I = new('icons/mob/blob.dmi', "blob")
-	if(overmind)
-		I.color = overmind.blob_reagent_datum.color
-	add_overlay(I)
```
todo: undo this since it makes the blob core overlay get the blob color and we don't want that. Same for node.

ChangelingRain commented on code/game/gamemodes/miniantags/revenant/revenant_blight.dm:
```diff
@@ -56,12 +53,8 @@
-					affected_mob.dna.species.handle_hair(affected_mob,"#1d2953")
-					affected_mob.visible_message("<span class='warning'>[affected_mob] looks terrifyingly gaunt...</span>", "<span class='revennotice'>You suddenly feel like your skin is <i>wrong</i>...</span>")
-					old_color = affected_mob.color
-					affected_mob.color = "#1d2953"
```
I'm not sure when this got in, but it shouldn't have actually been in here; the only affected color should have been the human's body itself, rather than the entire color.

ChangelingRain commented on code/game/gamemodes/miniantags/revenant/revenant_blight.dm:
```diff
@@ -56,12 +53,8 @@
-					affected_mob.color = "#1d2953"
-					addtimer(src, "cure", 100, FALSE)
+				affected_mob.visible_message("<span class='warning'>[affected_mob] looks terrifyingly gaunt...</span>", "<span class='revennotice'>You suddenly feel like your skin is <i>wrong</i>...</span>")
+				affected_mob.add_atom_colour("#1d2953", TEMPORARY_COLOUR_PRIORITY)
```
As such, this shouldn't be using the color system, as that behavior was wrong/produces the wrong visual effects.

phil235 commented on code/game/gamemodes/clock_cult/clock_scripture.dm:
```diff
@@ -365,8 +365,9 @@ Judgement: 10 servants, 100 CV, and any existing AIs are converted or destroyed
-	invoker.color = "#AF0AAF"
-	animate(invoker, color = initial(invoker.color), time = flee_time+grace_period)
+	var/invoker_old_color = invoker.color
+	invoker.add_atom_colour("#AF0AAF", TEMPORARY_COLOUR_PRIORITY)
```
todo: remove this line

ChangelingRain commented on code/game/gamemodes/nuclear/nuclearbomb.dm:
```diff
@@ -515,16 +515,12 @@ This is here to make the tiles around the station mininuke change when it's arme
-			user.color = "#00FF00"
-		else
-			user.color = "#FF0000"
+		user.add_atom_colour(RANDOM_COLOUR, ADMIN_COLOUR_PRIORITY)
```
This should be flickering between red and green, which appears to have been accidentally reverted?

phil235 replied:
oops, yeah, dunno what happed here.

ChangelingRain commented on code/game/machinery/shieldgen.dm:
```diff
@@ -65,6 +65,7 @@
 	name = "Invoker's Shield"
 	desc = "A weak shield summoned by cultists to protect them while they carry out delicate rituals"
 	color = "red"
+	atom_colours = list("", "", "", "red")
```
This isn't your fault but I really wish people would use actual colors rather than, uh, whatever color code this turns out to be?

RemieRichards replied:
I think red is just your standard 255,0,0,255 / #FF0000

phil235 commented on code/game/objects/effects/decals/crayon.dm:
```diff
@@ -37,14 +38,12 @@
 		qdel(src)
 
 	var/area/territory = get_area(location)
-	var/color
```
let's not use the same name as an actual variable of the src.

ChangelingRain commented on code/game/objects/items/weapons/weaponry.dm:
```diff
@@ -174,43 +174,43 @@ var/highlander_claymores = 0
 			user << "<span class='userdanger'>ANOTHER ONE FALLS TO YOUR BLOWS. ANOTHER WEAKLING UNFIT TO LIVE.</span>"
 			new_name = "gore-stained claymore"
-			color = rgb(255, 95, 95)
+			add_atom_colour(rgb(255, 95, 95), ADMIN_COLOUR_PRIORITY)
```
Shouldn't all of these be FIXED_COLOR_PRIORITY, given that they're inherent to the claymore?

phil235 replied:
I figured we want to see the claymore color all the time, even if it has paint on it, and since it's a rare magical admin event I didn't see any issue with using the admin colour priority.

phil235 commented on code/game/gamemodes/miniantags/revenant/revenant_abilities.dm:
```diff
@@ -343,16 +343,6 @@
-						if(H && H.dna && H.dna.species)
-							H.dna.species.handle_mutant_bodyparts(H)
-							H.dna.species.handle_hair(H)
-							H.color = old_color
```
this is conflicting with the blight virus below, if we keep this short temporary effect it will remove the blight virus' color when in stage 5 even if not cured yet. Another solution would be to put the two on different priorities but I don't want blight on admin priority or the temp effect on washable priority, and I don't to add a fifth priority just for this.

phil235 commented on code/game/gamemodes/miniantags/revenant/revenant_blight.dm:
```diff
@@ -56,12 +53,8 @@
 				PoolOrNew(/obj/effect/overlay/temp/revenant, affected_mob.loc)
-				if(affected_mob.dna && affected_mob.dna.species)
-					affected_mob.dna.species.handle_mutant_bodyparts(affected_mob,"#1d2953")
-					affected_mob.dna.species.handle_hair(affected_mob,"#1d2953")
```
I don't see the point of changing our hair color when we're giving the mob a color value that will color everything.

phil235 commented on code/game/objects/effects/effect_system/effects_smoke.dm:
```diff
@@ -297,16 +299,16 @@
 
 /datum/effect_system/smoke_spread/chem/start()
-	var/color = mix_color_from_reagents(chemholder.reagents.reagent_list)
+	var/mixcolor = mix_color_from_reagents(chemholder.reagents.reagent_list)
```
same shit

phil235 commented on code/game/objects/items/weapons/storage/toolbox.dm:
```diff
@@ -55,17 +55,17 @@
 /obj/item/weapon/storage/toolbox/electrical/New()
 	..()
-	var/color = pick("red","yellow","green","blue","pink","orange","cyan","white")
+	var/pickedcolor = pick("red","yellow","green","blue","pink","orange","cyan","white")
```
and again

phil235 commented on code/game/objects/items/weapons/storage/toolbox.dm:
```diff
@@ -93,12 +93,12 @@
 /obj/item/weapon/storage/toolbox/drone/New()
 	..()
-	var/color = pick("red","yellow","green","blue","pink","orange","cyan","white")
+	var/pickedcolor = pick("red","yellow","green","blue","pink","orange","cyan","white")
```
and again

phil235 commented on code/game/turfs/simulated/floor/misc_floor.dm:
```diff
@@ -159,3 +159,50 @@
 	floor_tile = /obj/item/stack/tile/sepia
 
+
+
```
Moved here because we don't define turf outside the turf folder please.

phil235 commented on code/modules/games/cards.dm:
```diff
@@ -19,17 +19,19 @@
 	. = ..()
 
-	var/color
+	var/cardcolor
```
and again

phil235 commented on code/modules/holiday/easter.dm:
```diff
@@ -111,9 +111,9 @@
 
 /obj/item/weapon/reagent_containers/food/snacks/egg/loaded/New()
 	..()
-	var/color = pick("blue","green","mime","orange","purple","rainbow","red","yellow")
```
jskdfjksqgsdg tshf hsdfs

ChangelingRain commented:
oh god my review slipped in the middle of yours

RemieRichards commented on code/game/atoms.dm:
```diff
@@ -436,8 +436,70 @@ var/list/blood_splatter_icons = list()
+*/
+/atom/proc/update_atom_colour()
+	if(!atom_colours)
+		atom_colours = list("", "", "", "")
```
either:
Add a define for the number of priorities (eg: ```#define COLOUR_PRIORITY_AMOUNT 4```) and then do 
```
atom_colours = list()
atom_colours.len = COLOUR_PRIORITY_AMOUNT
```

or add a global list of ```list("", "", "", "")``` to ```.Copy()``` from in resets, as it stands you use the list ```list("","","","")``` in atleast 3 places, and that's 3 places it's going to break/need updating if another priority is added.

AnturK commented:
Is the doubling of definitions a technical issue or choice ?

phil235 commented on code/game/objects/effects/decals/crayon.dm:
```diff
@@ -25,6 +25,7 @@
 		src.transform = M
 
 	color = main
+	atom_colours[FIXED_COLOUR_PRIORITY] = main
```
could cause issue because the atom_colours list var is still null at that point.

phil235 commented on code/game/objects/structures/beds_chairs/chair.dm:
```diff
@@ -128,6 +128,7 @@
 	desc = "It looks comfy.\n<span class='notice'>Alt-click to rotate it clockwise.</span>"
 	icon_state = "comfychair"
 	color = rgb(255,255,255)
+	atom_colours = list("", "", "", rgb(255,255,255))
```
shoudld probably do this in comfy/New()

phil235 commented:
>Is the doubling of definitions a technical issue or choice ?

I don't understand your question. what are you refering to?

AnturK commented:
Having to define both color var and the composite list. I suppose you'd have to shunt it up somewhere high up the type tree if it was in New

MrStonedOne commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1307,10 +1307,10 @@ proc/pick_closest_path(value)
 	C.color = flash_color
 	spawn(0)
-		animate(C, color = initial(C.color), time = flash_time)
+		animate(C, color = old_color, time = flash_time)
```
why is this in a spawn, i'm not seeing a sleep.

animate doesn't block.

This spawn seems unneeded.

phil235 committed:
fixing stuff about atom_colours.

Shadowlight213 commented:
Color vs colour III, the memes strike back

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into Color_washing

# Conflicts:
#	code/game/gamemodes/clock_cult/clock_scripture.dm
#	code/game/gamemodes/clock_cult/clock_structures.dm

phil235 committed:
Fixes merge conflict stuff.
Readds the lines about coloring hairs and mutant parts in revblight (even though they're no longer needed, but it's a reminder that the color change is a trick and we need to replace it with a proper forced limb coloring like it used to).

ChangelingRain commented on code/game/objects/effects/effect_system/effects_smoke.dm:
```diff
@@ -150,6 +150,7 @@
 /obj/effect/particle_effect/smoke/freezing
 	name = "nanofrost smoke"
 	color = "#B2FFFF"
+	atom_colours= list("", "", "", "#B2FFFF")
```
Find+replace probably missed this

ChangelingRain commented on code/modules/events/wizard/greentext.dm:
```diff
@@ -87,7 +87,7 @@
 		if(M in color_altered_mobs)
 			message += " and you're finally able to forgive yourself"
-			M.color = initial(M.color)
+			M.remove_atom_colour(ADMIN_COLOUR_PRIORITY)
```
This should make sure the admin color is one of the two colors this can cause, just in case something else gets them/an admin fucks around

phil235 committed:
Tiny fixes and suggested changes.
Blight ability will still color your hair, and will be cleared when cured.
